pkg/logs/sources/config_source.go

```diff
@@ -0,0 +1,100 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package sources
+
+import (
+	"os"
+	"sync"
+
+	logsConfig "github.com/DataDog/datadog-agent/comp/logs/agent/config"
+)
+
+// ConfigSources receives file paths to log configs and creates sources. The sources are added to a channel and read by the launcher.
+// This class implements the SourceProvider interface
+type ConfigSources struct {
+	addedByType map[string][]chan *LogSource
+}
+
+var (
+	instance *ConfigSources
+	once     sync.Once
+)
+
+// GetInstance provides a singleton instance of ConfigSources.
+func GetInstance() *ConfigSources {
+	once.Do(func() {
+		instance = &ConfigSources{
+			addedByType: make(map[string][]chan *LogSource),
+		}
+	})
+	return instance
+}
+
+// AddFileSource gets a file from a file path and adds it as a source.
+func (s *ConfigSources) AddFileSource(path string) error {
+	wd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	absolutePath := wd + "/" + path
+	data, err := os.ReadFile(absolutePath)
+	if err != nil {
+		return err
+	}
+	logsConfig, err := logsConfig.ParseYAML(data)
+	if err != nil {
+		return err
+	}
+	configSource := GetInstance()
+	for _, cfg := range logsConfig {
+		if cfg.TailingMode == "" {
+			cfg.TailingMode = "beginning"
+		}
+		source := NewLogSource(cfg.Name, cfg)
+		configSource.AddSource(source)
+	}
+
+	return nil
+}
+
+// AddSource adds a new source.
+// All of the subscribers registered for this source's type (src.Config.Type) will be
+// notified.
+func (s *ConfigSources) AddSource(source *LogSource) {
```

Comment:
The complexity here can be reduced. The reason the regular source provider is complex is because we cannot guarantee ordering of when the launcher starts vs when a source becomes available. 

But for this we can: 
- if `AddSource` is always called before `launcher.start` (which I think you would need to change) this code could simply be:

in `AddSource` store the source in a map of list of sources (by type)
in `SubscribeForType` just return a new channel with the sourcesfrom the map written to it. 

IMO it would make this code easier to reason about.